pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+//#[cfg(feature = "runtime-benchmarks")]
+//mod benchmarking;
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{Currency, ExistenceRequirement, ReservableCurrency},
+	transactional, PalletId,
+};
+use frame_system::pallet_prelude::*;
+use num_traits::{NumAssignOps, NumOps, Signed};
+use sp_runtime::{
+	traits::{AccountIdConversion, BlockNumberProvider, Saturating, Zero},
+	FixedPointNumber, FixedPointOperand, TokenError,
+};
+
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, RuntimeDebug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct EpochDetails<BlockNumber, Balance> {
+	ends_on: BlockNumber,
+	total_reward: Balance,
+}
+
+/// Type used to initialize the first epoch with the correct block number
+pub struct FirstEpochDetails<P>(std::marker::PhantomData<P>);
+impl<P, N, B: Zero> Get<EpochDetails<N, B>> for FirstEpochDetails<P>
+where
+	P: BlockNumberProvider<BlockNumber = N>,
+{
+	fn get() -> EpochDetails<N, B> {
+		EpochDetails {
+			ends_on: P::current_block_number(),
+			total_reward: Zero::zero(),
+		}
+	}
+}
+
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, RuntimeDebug, Default)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct GroupDetails<Balance, Rate> {
+	total_staked: Balance,
+	reward_per_token: Rate,
+}
+
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, RuntimeDebug, Default)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct StakedDetails<Balance, SignedBalance> {
+	amount: Balance,
+	reward_tally: SignedBalance,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	pub type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		#[pallet::constant]
+		type BlockPerEpoch: Get<Self::BlockNumber>;
+
+		type Currency: ReservableCurrency<Self::AccountId>;
+
+		type SignedBalance: From<BalanceOf<Self>>
```

Comment:
What is the purpose of a `Signed` Balance?

Reply:
It's a type that is expected to behave as a Balance but can be negative in some cases. It's used to hold the `reward_tally`.